File: src/algorithms/trees/advanced/huffman-coding-tree/__tests__/HuffmanCodingTree_test.java
// javac *.java && java -ea HuffmanCodingTree_test
import java.util.Map;

public class HuffmanCodingTree_test {
    public static void main(String[] args) {
        HuffmanCodingTree hct = new HuffmanCodingTree();

        char[] chars = {'a', 'b', 'c', 'd', 'e', 'f'};
        int[] freqs = {5, 9, 12, 13, 16, 45};

        // test: produces encodings for all characters
        Map<Character, String> result = hct.huffmanCodingTree(chars, freqs);
        for (char ch : chars) {
            assert result.containsKey(ch) : "Missing encoding for " + ch;
            assert result.get(ch) != null && !result.get(ch).isEmpty() : "Empty encoding for " + ch;
        }

        // test: produces valid binary strings
        for (Map.Entry<Character, String> entry : result.entrySet()) {
            assert entry.getValue().matches("[01]+") : "Invalid encoding: " + entry.getValue();
        }

        // test: most frequent character gets shortest code
        int fLen = result.get('f').length();
        for (char ch : chars) {
            if (ch != 'f') {
                assert fLen <= result.get(ch).length() : "f should have shortest code";
            }
        }

        // test: all codes are prefix-free
        String[] codes = result.values().toArray(new String[0]);
        for (int idxA = 0; idxA < codes.length; idxA++) {
            for (int idxB = 0; idxB < codes.length; idxB++) {
                if (idxA != idxB) {
                    assert !(codes[idxA].startsWith(codes[idxB]) && !codes[idxA].equals(codes[idxB]))
                        : "Prefix conflict: " + codes[idxA] + " starts with " + codes[idxB];
                }
            }
        }

        // test: single character
        char[] singleChar = {'x'};
        int[] singleFreq = {10};
        Map<Character, String> singleResult = hct.huffmanCodingTree(singleChar, singleFreq);
        assert singleResult.get('x').equals("0") : "Single char encoding should be 0";

        System.out.println("All tests passed!");
    }
}

File: src/algorithms/trees/advanced/huffman-coding-tree/__tests__/huffman-coding-tree_test.go
package main

import (
	"strings"
	"testing"
)

var defaultHuffmanFreqs = []CharFreq{
	{Char: 'a', Freq: 5},
	{Char: 'b', Freq: 9},
	{Char: 'c', Freq: 12},
	{Char: 'd', Freq: 13},
	{Char: 'e', Freq: 16},
	{Char: 'f', Freq: 45},
}

func isValidBinary(code string) bool {
	for _, bit := range code {
		if bit != '0' && bit != '1' {
			return false
		}
	}
	return len(code) > 0
}

func TestHuffmanProducesEncodingsForAll(t *testing.T) {
	result := huffmanCodingTree(defaultHuffmanFreqs)
	for _, item := range defaultHuffmanFreqs {
		if _, ok := result[item.Char]; !ok {
			t.Errorf("missing encoding for %c", item.Char)
		}
	}
}

func TestHuffmanValidBinaryStrings(t *testing.T) {
	result := huffmanCodingTree(defaultHuffmanFreqs)
	for ch, encoding := range result {
		if !isValidBinary(encoding) {
			t.Errorf("invalid encoding for %c: %q", ch, encoding)
		}
	}
}

func TestHuffmanMostFrequentGetsShortest(t *testing.T) {
	result := huffmanCodingTree(defaultHuffmanFreqs)
	fLen := len(result['f'])
	for _, item := range defaultHuffmanFreqs {
		if item.Char != 'f' {
			if fLen > len(result[item.Char]) {
				t.Errorf("f should have shortest code, but %c has shorter", item.Char)
			}
		}
	}
}

func TestHuffmanPrefixFree(t *testing.T) {
	result := huffmanCodingTree(defaultHuffmanFreqs)
	codes := make([]string, 0)
	for _, enc := range result {
		codes = append(codes, enc)
	}
	for idxA, codeA := range codes {
		for idxB, codeB := range codes {
			if idxA != idxB && strings.HasPrefix(codeA, codeB) && codeA != codeB {
				t.Errorf("code %q is a prefix of %q", codeB, codeA)
			}
		}
	}
}

func TestHuffmanSingleCharacter(t *testing.T) {
	result := huffmanCodingTree([]CharFreq{{Char: 'x', Freq: 10}})
	if result['x'] != "0" {
		t.Errorf("expected '0', got %q", result['x'])
	}
}
